src/SAWScript/SolverCache.hs

```diff
@@ -0,0 +1,483 @@
+--    depend on any execution-specific 'VarIndex'es.
+-- 2. After calling 'scWriteExternal', all 'LocalName's in 'Pi' and 'Lam'
+--    constructors are removed. This ensures that two terms which are alpha
+--    equivalent are given the same hash.
+mkSolverCacheKey :: SharedContext -> SolverBackendVersions ->
+                    [SolverBackendOption] -> SATQuery -> IO SolverCacheKey
+mkSolverCacheKey sc vs opts satq = do
+  body <- satQueryAsPropTerm sc satq
+  let ecs = M.keys (satVariables satq) ++
+            filter (\ec -> ecVarIndex ec `elem` satUninterp satq)
+                   (getAllExts body)
+  tm <- scGeneralizeExts sc ecs body
+  let str_prefix = [ showBackendVersionsWithOptions "\n" vs opts
+                   , "satVariables " ++ show (M.size (satVariables satq))
+                   , "satUninterp "  ++ show (length (satUninterp  satq)) ]
+      str_to_hash = unlines str_prefix ++ anonLocalNames (scWriteExternal tm)
+  return $ SolverCacheKey vs opts $ SHA256.hash $ encodeUtf8 $ T.pack $ str_to_hash
+  where anonLocalNames = unlines . map (unwords . go . words) . lines
+        go (x:y:_:xs) | y `elem` ["Pi", "Lam"] = x:y:"_":xs
+        go xs = xs
+
+
+-- Solver Cache Values ---------------------------------------------------------
+
+-- | The value type for 'SolverCache': a 'Set' of 'SolverBackendVersion's, a
+-- 'String' representing the solver used, and an optional list of
+-- counterexamples, represented as pairs of indexes into the list of
+-- 'satVariables' of an associated 'SATQuery'
+data SolverCacheValue =
+  SolverCacheValue
+  { solverCacheValueVersions   :: SolverBackendVersions
+  , solverCacheValueOptions    :: [SolverBackendOption]
+  , solverCacheValueSolverName :: String
+  , solverCacheValueCEXs       :: Maybe [(Int, FirstOrderValue)]
+  } deriving Eq
+
+instance FromJSON SolverCacheValue where
+  parseJSON = JSON.withObject "SolverCacheValue" $ \v -> do
+    vs      <- v .:  "vs"
+    opts    <- v .:? "opts"
+    nm      <- v .:  "nm"
+    mb_cexs <- v .:? "cexs"
+    return $ SolverCacheValue vs (fromMaybe [] opts) nm mb_cexs
+
+instance ToJSON SolverCacheValue where
+  toJSON (SolverCacheValue vs opts nm mb_cexs) = JSON.object $
+    ["vs" .= vs] ++ (if null opts then [] else ["opts" .= opts]) ++
+    ["nm" .= nm] ++ maybe [] (\cexs -> ["cexs" .= cexs]) mb_cexs
+  toEncoding (SolverCacheValue vs opts nm mb_cexs) = JSON.pairs $
+    "vs" .= vs <> (if null opts then mempty else "opts" .= opts) <>
+    "nm" .= nm <> maybe mempty (\cexs -> "cexs" .= cexs) mb_cexs
+
+-- | Convert the result of a solver call on the given 'SATQuery' to a
+-- 'SolverCacheValue'
+toSolverCacheValue :: SolverBackendVersions -> [SolverBackendOption] ->
+                      SATQuery -> (Maybe CEX, String) -> Maybe SolverCacheValue
+toSolverCacheValue vs opts satq (cexs, solver_name) =
+  fmap (SolverCacheValue vs opts solver_name)
+       (mapM (mapM (firstM (`elemIndex` ecs))) cexs)
```

Comment:
`fmap ... (mapM (mapM ...) ...)`... there is so much nested mapping going on that I struggle to understand what is going on here. I think it would be worth breaking this up into multiple lines (with `do`-notation) to make it easier to understand at a glance.

Reply:
I wasn't sure how to rewrite this in `do`-notation, but instead I tried pulling the confusing jumble of `mapM`s etc. out as a subfunction with an explicit type. Is something like the below more readable now?
```haskell
  SolverCacheValue vs opts solver_name <$> firstsMaybeM (`elemIndex` ecs) cexs
  where firstsMaybeM :: Monad m => (a -> m b) -> Maybe [(a, c)] -> m (Maybe [(b, c)])
        firstsMaybeM = mapM . mapM . firstM
```
I also made the analogous change to `fromSolverCacheValue` and it's definitely better that what it was before: `(fmap (fmap (first (ecs !!))) cexs, solver_name)`.
```haskell
  (firstsMaybe (ecs !!) cexs, solver_name)
  where firstsMaybe :: (a -> b) -> Maybe [(a, c)] -> Maybe [(b, c)]
        firstsMaybe = fmap . fmap . first
```

Reply:
Yes, that's much clearer now.